target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+	if (IS_ERR(rtcl_ccu))
+		return -ENOMEM;
+
+	rtcl_ccu->soc = soc;
+	rtcl_ccu->dram.type = RTL_MC_MCR_DRAMTYPE(read_soc(RTL_MC_MCR));
+	rtcl_ccu->dram.buswidth = RTL_MC_DCR_BUSWIDTH(read_soc(RTL_MC_DCR));
+	rtcl_ccu->np = np;
+	spin_lock_init(&rtcl_ccu->lock);
+
+	return 0;
+}
+
+void rtcl_init_rates(int clk_idx)
+{
+	u32 vals[2];
+	int ret;
+	const struct rtcl_round_set *round = &rtcl_round_set[rtcl_ccu->soc][clk_idx];
+	struct rtcl_clk *rclk = &rtcl_ccu->clks[clk_idx];
+	struct clk *clk = clk_get(NULL, rtcl_clk_info[clk_idx].name);
+/*
+ * Determine startup clock rates by default framework. For min/max values we will mix the DT
+ * settings with the internal tables. DDR3 design allows a minimum of 303 MHz. To get below that
+ * barrier we must switch the memory controller into a DLL off/disable mode. Keep the complexity
+ * low and avoid this stuff for now. The 839x is a little bit picky.
+ */
+	rclk->startup = clk_get_rate(clk);
+
+	ret = of_property_read_u32_array(rtcl_ccu->np, rtcl_clk_info[clk_idx].dtrange, vals, 2);
+	if (ret || (vals[1] < vals[0]))
+		vals[0] = vals[1] = rtcl_ccu->clks[clk_idx].startup;
+
+	rclk->min = max_t(unsigned long, vals[0], round->min);
+	rclk->max = min_t(unsigned long, vals[1], round->max);
+
+	if (clk_idx == CLK_MEM && rtcl_ccu->dram.type == MEM_DDR3) {
+		switch (rtcl_ccu->soc) {
+		case SOC_RTL838X:
+			rclk->min = max(rclk->min, 300000000ul);
+			break;
+		case SOC_RTL839X:
+			rclk->min = max(rclk->min, 325000000ul);
+			break;
+		}
+	}
+
+	rclk->min = clk_round_rate(clk, 0ul);
+	rclk->max = clk_round_rate(clk, ~0ul);
+	clk_put(clk);
+
+	rtcl_ccu->dynamic |= rclk->min < rclk->max;
+}
+
+int rtcl_register_clkhw(int clk_idx)
+{
+	struct clk_parent_data parent_data = { };
+	struct clk_init_data hw_init = { };
+	struct rtcl_clk *rclk = &rtcl_ccu->clks[clk_idx];
+	int ret;
+
+	rclk->hw.init = &hw_init;
+	rclk->idx = clk_idx;
+
+	parent_data.fw_name = rtcl_clk_info[clk_idx].parent_name;
+
+	hw_init.name = rtcl_clk_info[clk_idx].name;
+	hw_init.ops = &rtcl_clk_ops;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
+
+	ret = of_clk_hw_register(rtcl_ccu->np, &rclk->hw);
+	if (ret)
+		return ret;
+
+	clk_hw_register_clkdev(&rclk->hw, rtcl_clk_info[clk_idx].name, NULL);
+	rtcl_init_rates(clk_idx);
+	return 0;
+}
+
+static struct clk_hw *rtcl_get_clkhw(struct of_phandle_args *clkspec, void *prv)
+{
+	unsigned int idx = clkspec->args[0];
+
+	if (idx >= CLK_COUNT) {
+		pr_err("%s: Invalid index %u\n", __func__, idx);
+		return ERR_PTR(-EINVAL);
+	}
+
+	return &rtcl_ccu->clks[idx].hw;
+}
+
+static int rtcl_ccu_register_clocks(void)
+{
+	int clk_idx, ret;
+
+	for (clk_idx = 0; clk_idx < CLK_COUNT; ++clk_idx) {
+		ret = rtcl_register_clkhw(clk_idx);
+		if (ret) {
+			pr_err("%s: Couldn't register %s clock\n",
+				__func__, rtcl_clk_info[clk_idx].display_name);
+			goto err_hw_unregister;
+		}
+	}
+
+	ret = of_clk_add_hw_provider(rtcl_ccu->np, rtcl_get_clkhw, rtcl_ccu);
+	if (!ret)
+		return 0;
+
```

Comment:
does it make sense to swap the two? if (ret) -> print error, goto error; and then 'return 0; for success? almost reads like we return 0 on error ... reads a little bit as an anti-pattern.

Reply:
If you like